[JIT] Print out CU address in `ClassType::repr_str()`

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50194 [JIT] Print out CU address in `ClassType::repr_str()`**

**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

Differential Revision: [D25933082](https://our.internmc.facebook.com/intern/diff/D25933082)

Commit:
[JIT] Print out CU address in `ClassType::repr_str()`

**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling.

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

[ghstack-poisoned]

Commit:
Update on "[JIT] Print out CU address in `ClassType::repr_str()`"


**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

[ghstack-poisoned]

SplitInfinity commented:
Hm, it's probably a better idea to print out the type as `ClassType (compilation unit: 0x2398579)` instead of `ClassType (0x2398579)` which is probably just as confusing as `ClassType`.

Commit:
Update on "[JIT] Print out CU address in `ClassType::repr_str()`"


**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

[ghstack-poisoned]

SplitInfinity commented:
For clarity, the original issue was opened by someone who used the same TorchScript class in two different modules, saved them into two different archives, and then tried to pass an instance of that class generated by one module to the other. This didn't work because the one TorchScript class corresponds to two different JIT type objects (since the modules were serialized separately), and those objects are not equal (`ClassType::operator==`) to each other because their compilation units don't match. I think the fix here is to save both into the same archive, but I think the error message (e.g. `expected type T but got type T`) can be improved, and that's what this PR is for.

gmagogsfm commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1869,7 +1869,13 @@ struct TORCH_API ClassType : public NamedType {
+
+  std::string repr_str() const override {
+    std::stringstream ss;
+    ss << str() << "(" << compilation_unit().get() << ")";
```
Nit:

Having a memory address trailing a type is puzzling too : )

How about this?
```
ss << str() << "(of Python compilation unit at: " << compilation_unit().get() << ")";
```

gmagogsfm replied:
Interesting idea. They could have names. In this case, it would be inferred from which imported module they come from. But I guess this is a slightly bigger change that has usability implications and warrants a discussion among team. What do you think?

SplitInfinity replied:
Yeah, makes sense, I can file and issue for it.

gmagogsfm commented on test/cpp/jit/test_class_type.cpp:
```diff
@@ -55,5 +56,51 @@ TEST(ClassTypeTest, AddRemoveConstant) {
+  push(stack, obj._ivalue());
+  try {
+    method(stack, {});
+  } catch (const std::exception& e) {
```
This test would pass when no exception is thrown, should we assert after this try-catch pair and do early-return in catch branch?

SplitInfinity replied:
Yes, good idea.

Commit:
Update on "[JIT] Print out CU address in `ClassType::repr_str()`"


**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

[ghstack-poisoned]

Commit:
Update on "[JIT] Print out CU address in `ClassType::repr_str()`"


**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

[ghstack-poisoned]

SplitInfinity commented:
Ping, I have made the requested changes.

gmagogsfm commented on test/cpp/jit/test_class_type.cpp:
```diff
@@ -55,5 +56,56 @@ TEST(ClassTypeTest, AddRemoveConstant) {
+    // Check that the exception contains the address of the compilation unit
+    // in addition to the ClassType's name.
+    testing::FileCheck()
+        .check("foo(0x")
```
Does this test still pass after changing error message? It should be 

```
.check("foo(of python")
.check_same("0x")
```

SplitInfinity replied:
Turns out I did not make the requested changes. 🙃 🔫

Commit:
Update on "[JIT] Print out CU address in `ClassType::repr_str()`"


**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

[ghstack-poisoned]

Commit:
Update on "[JIT] Print out CU address in `ClassType::repr_str()`"


**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

[ghstack-poisoned]

Commit:
Update on "[JIT] Print out CU address in `ClassType::repr_str()`"


**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

[ghstack-poisoned]

Commit:
Update on "[JIT] Print out CU address in `ClassType::repr_str()`"


**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

Differential Revision: [D25933082](https://our.internmc.facebook.com/intern/diff/D25933082)

[ghstack-poisoned]

Commit:
Update on "[JIT] Print out CU address in `ClassType::repr_str()`"


**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

Differential Revision: [D25933082](https://our.internmc.facebook.com/intern/diff/D25933082)

[ghstack-poisoned]

Commit:
Update on "[JIT] Print out CU address in `ClassType::repr_str()`"


**Summary**
`ClassType::repr_str()` prints out only the name of a `ClassType`, which
is not always enough to disambiguate it. In some situations, two
`ClassTypes` are compared and do not match despite having identical
names because they are in separate compilation units. In such cases, the
error message can seem nonsensical (e.g. `expected type T but found type
T`). This commit modifies `ClassType::repr_str()` so that it prints out
the address of the type's compilation unit to make these messages less
puzzling (e.g. `expected type T (0x239023) but found type T (0x230223)`).

**Test Plan**
This commit adds a unit test, `ClassTypeTest.IdenticalTypesDifferentCus`
that reproduces this situation.

**Fixes**
This commit fixes #46212.

Differential Revision: [D25933082](https://our.internmc.facebook.com/intern/diff/D25933082)

[ghstack-poisoned]
